lib/shopify_api/bulk/query.ex

```diff
@@ -0,0 +1,195 @@
+         {:ok, resp} <- ShopifyAPI.graphql_request(token, bulk_query, 10),
+         :ok <- handle_errors(resp),
+         bulk_query_id <- get_in(resp.response, ["bulkOperationRunQuery", "bulkOperation", "id"]),
+         {:ok, url} <- poll(token, bulk_query_id, opts[:polling_rate], opts[:max_poll_count]) do
+      {:ok, url}
+    else
+      {:error, :timeout, bulk_id} ->
+        maybe_cancel(opts[:auto_cancel], token, bulk_id)
+        {:error, @polling_timeout_message}
+
+      error ->
+        error
+    end
+  end
+
+  # Shopify returns a single newline which gets stripped and we are left with garbage,
+  # handle it nicely here.
+  def fetch(nil), do: {:ok, ""}
+
+  def fetch({:error, _} = error), do: error
+  def fetch({:ok, url}), do: fetch(url)
+
+  def fetch(url) when is_binary(url) do
+    url
+    |> HTTPoison.get()
+    |> case do
+      {:ok, %{body: jsonl}} -> {:ok, jsonl}
+      error -> error
+    end
+  end
+
+  def parse_response!(""), do: []
+  def parse_response!({:ok, jsonl}), do: parse_response!(jsonl)
+  def parse_response!({:error, msg}), do: raise(ShopifyAPI.Bulk.QueryError, msg)
+
+  def parse_response!(jsonl) when is_binary(jsonl) do
+    jsonl
+    |> String.split("\n", trim: true)
+    |> Enum.map(&ShopifyAPI.JSONSerializer.decode!/1)
+  end
+
+  @spec status(AuthToken.t()) :: {:ok, status_response()} | {:error, any()}
+  def status(%AuthToken{} = token) do
+    token
+    |> ShopifyAPI.graphql_request(@polling_query, 1)
+    |> case do
+      {:ok, %{response: %{"currentBulkOperation" => response}}} ->
+        {:ok, response}
+
+      {:error, _} = error ->
+        error
+    end
+  end
+
+  defp handle_errors(resp) do
+    errors = get_in(resp.response, ["bulkOperationRunQuery", "userErrors"])
+
+    case Enum.empty?(errors) do
+      true -> :ok
+      false -> {:error, fetch_first_error(errors)}
+    end
+  end
+
+  defp bulk_query_string(query) do
+    query_string = ShopifyAPI.JSONSerializer.encode!(query)
+
+    """
+    mutation {
+      bulkOperationRunQuery(
+        query: #{query_string}
+      ) {
+        bulkOperation {
+          id
+          status
+        }
+        userErrors {
+          field
+          message
+        }
+      }
+    }
+    """
+  end
+
+  defp fetch_first_error(errors) do
+    errors
+    |> List.first()
+    |> Map.get("message")
+  end
+
+  defp poll(token, bulk_query_id, polling_rate, max_poll, depth \\ 0)
+
+  defp poll(token, bulk_query_id, _, max_poll, depth) when max_poll == depth,
+    do: {:error, :timeout, bulk_query_id}
+
+  defp poll(token, bulk_query_id, polling_rate, max_poll, depth) do
+    Process.sleep(polling_rate)
+
+    case status(token) do
+      {:ok, %{"status" => "COMPLETED", "url" => url} = _response} -> {:ok, url}
+      _ -> poll(token, bulk_query_id, polling_rate, max_poll, depth + 1)
+    end
+  end
+
+  defp maybe_cancel(false, _, _), do: nil
+  defp maybe_cancel(true, token, bid), do: token |> cancel(bid) |> poll_till_cancel(token, bid)
```

Comment:
🔥This shouldn't be passing `bid` as an argument to `poll_till_cancel` unless we want to start REALLY BIG max_poll values.

Reply:
Hmm, the compile warning this causes is odd:
`the first 3 default arguments in poll_till_cancel/4 are never used`

When it's actually the first default argument, but the func's third arg.